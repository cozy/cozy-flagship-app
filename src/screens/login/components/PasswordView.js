import React, {useEffect, useState} from 'react'
import {Button, View, Text, TextInput} from 'react-native'
import Minilog from '@cozy/minilog'

import {queryResultToCrypto} from '../../../components/webviews/CryptoWebView/cryptoObservable/cryptoObservable'

const log = Minilog('PasswordView')

const ITERATION_NUMBER = 100_000

const getSaltForInstance = instance => {
  const domain = instance.split(':')[0]
  const salt = `me@${domain}`

  return salt
}

/**
 * Hash password data by calling CryptioWebView cryptography methods
 * @param {PasswordData} passwordData - the password data to hash
 * @param {string} instance - the Cozy instance used to generate the salt
 * @returns {LoginData} login data containing hashed password and encryption keys
 */
const doHashPassword = async (passwordData, fqdn) => {
  log.debug('Start hashing password')
  try {
    const {password, hint} = passwordData

    const salt = getSaltForInstance(fqdn)

    const result = await queryResultToCrypto('computePass', {
      pass: password,
      salt: salt,
      iterations: ITERATION_NUMBER,
    })

    const {iterations, key, publicKey, privateKey, passwordHash} = result.param

    const loginData = {
      passwordHash,
      hint,
      iterations,
      key,
      publicKey,
      privateKey,
    }

    return loginData
  } catch (e) {
    log.error('Error while requesting cryptography result:', e)
    throw e
  }
}

/**
 * Show a password form that asks the user their password and hint
 * When the user validate their password, the password and the salt are sent back to parent
 * by calling `setPasswordData`
 *
 * @param {object} props
 * @param {setPasswordData} props.setPasswordData
 * @param {string} props.fqdn
 * @returns {import('react').ComponentClass}
 */
const PasswordForm = ({setPasswordData, fqdn}) => {
  const [password, onChangePassword] = useState('')
  const [hint, onChangeHint] = useState('')

  const onLogin = async () => {
    setPasswordData({
      password,
      hint,
    })
  }

  return (
    <>
      <Text>FQDN</Text>
      <Text>{fqdn}</Text>

      <Text>Password</Text>
      <TextInput onChangeText={onChangePassword} value={password} />

      <Text>Hint</Text>
      <TextInput onChangeText={onChangeHint} value={hint} />

      <Button onPress={onLogin} title="Login" accessibilityLabel="Login" />
    </>
  )
}

/**
 * Show a password view that asks the user their password and hint
 * When the user validate their password, the password is hashed and is send back to parent
 * with corresponding cryptographic keys by calling `setKeys`
 * If an error occurs, then `setError` is called
 *
 * @param {object} props
 * @param {string} props.fqdn
 * @param {setLoginDataCallback} props.setKeys
 * @param {setErrorCallback} props.setError
 * @param {ButtonInfo} props.cancelStep
 * @param {ButtonInfo} [props.cancelAll]
 * @returns {import('react').ComponentClass}
 */
export const PasswordView = ({
  fqdn,
  setKeys,
  setError,
  cancelStep,
  cancelAll = undefined,
  errorMessage,
}) => {
  const [passwordData, setPasswordData] = useState()

  useEffect(() => {
    if (passwordData) {
      doHashPassword(passwordData, fqdn)
        .then(result => {
          setKeys(result)
        })
        .catch(error => {
          setError('Impossible to hash the password', error)
        })
    }
  }, [passwordData, fqdn, setKeys, setError])

  return (
    <View>
      <PasswordForm setPasswordData={setPasswordData} fqdn={fqdn} />
      {errorMessage && <Text>{errorMessage}</Text>}
      <View style={{marginTop: 20}}>
        <Button
          onPress={() => cancelStep.callback()}
          title={cancelStep.title}
        />
      </View>
      {cancelAll && (
        <View style={{marginTop: 20}}>
          <Button
            onPress={() => cancelAll.callback()}
            title={cancelAll.title}
          />
        </View>
      )}
    </View>
  )
}
