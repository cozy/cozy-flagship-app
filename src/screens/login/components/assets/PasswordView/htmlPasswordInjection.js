/********************************************************************/
/* This code should reflect cozy-stack/assets/templates/login.html */
/********************************************************************/

import { encode } from 'html-entities'

import { handleErrorMessage } from './js/jsHandleErrorMessage'
import { loginJs } from './js/jsLogin'
import { passwordHelperJs } from './js/jsPasswordHelper'
import { passwordVisibilityJs } from './js/jsPasswordVisibility'
import { cirrusCss } from '../common/css/cssCirrus'
import { cozyBsCss } from '../common/css/cssCozyBs'
import { fontsCss } from '../common/css/cssFonts'
import { themeCss } from '../common/css/cssTheme'
import { cirrusJs } from '../common/js/jsCirrus'
import { readonlyJs } from '../common/js/jsReadonly'

const strBackButton = "Revenir à l'écran précédent"
const strPasswordField = 'Mot de passe'
const strPasswordHide = 'Cacher le mot de passe'
const strPasswordShow = 'Afficher le mot de passe'
const strForgotPassword = "J'ai oublié mon mot de passe"
const strSubmit = 'SE CONNECTER'

const locale = 'fr'

/**
 * Generate password specific HTML to provide to webview
 *
 * @param {string} title - The title of the page - for example the user's name as configured in the Cozy's settings
 * @param {string} fqdn - The subtitle of the page - for example the Cozy's fqdn
 * @param {string} instance - The Cozy's url, used to get avatar and fonts css
 * @returns {Element} HTML of Password form to inject inside Webview
 */
export const getHtml = (title, fqdn, instance) => {
  const avatarUrl = new URL(instance)
  avatarUrl.pathname = 'public/avatar'

  const avatarSvgUrl =
    'data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMTYiIGhlaWdodD0iMTYiIHZpZXdCb3g9IjAgMCAyNCAyNCIgZmlsbD0ibm9uZSIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIj4KPGcgdHJhbnNmb3JtPSJ0cmFuc2xhdGUoNCA0KSI+CjxwYXRoIGZpbGwtcnVsZT0iZXZlbm9kZCIgY2xpcC1ydWxlPSJldmVub2RkIiBkPSJNMTAuNDM2NSAxMC4wMDFDMTAuMzE5IDEwLjE4NCAxMC4wNTg1IDEwLjIzOSA5Ljg3ODk5IDEwLjExOUM5Ljg2MzQ5IDEwLjEwOTUgOS44NDg0OSAxMC4wOTkgOS44MzQ0OSAxMC4wODg1QzkuMzEwOTkgMTAuNDk5IDguNjY2NDkgMTAuNzIzIDcuOTk2NDkgMTAuNzIzQzcuMzI3OTkgMTAuNzIzIDYuNjg0OTkgMTAuNSA2LjE2MjQ5IDEwLjA5MUM2LjE0ODk5IDEwLjEwMDUgNi4xMzU0OSAxMC4xMTA1IDYuMTIxNDkgMTAuMTE5QzUuOTM2NDkgMTAuMjM5NSA1LjY3OTk5IDEwLjE4MiA1LjU2MzQ5IDEwLjAwMjVDNS40NDU0OSA5LjgyMDQ4IDUuNDk2OTkgOS41NzM5OCA1LjY3Nzk5IDkuNDUyNDhDNS44MzQ5OSA5LjM0ODk4IDUuODQ1OTkgOS4xNDA0OCA1Ljg0NTk5IDkuMTQwNDhDNS44NDk5OSA5LjAzMjQ5IDUuODk2OTkgOC45Mjk0OCA1Ljk3NTQ5IDguODU2OThDNi4wNTM0OSA4Ljc4NDQ4IDYuMTUxNDkgOC43NDI0OCA2LjI1OTQ5IDguNzUwOThDNi40NzU5OSA4Ljc1NTk4IDYuNjUxOTkgOC45Mzc5OCA2LjY0Nzk5IDkuMTU2OThDNi42NDc5OSA5LjE1OTk4IDYuNjQ3NDkgOS4yNjk0OCA2LjYwNzk5IDkuNDIxNDhDNy40MTU0OSAxMC4wOTIgOC41ODU5OSAxMC4wODk1IDkuMzkwNDkgOS40MTU5OEM5LjM2MDk5IDkuMjk5NDggOS4zNTM0OSA5LjIwNjQ4IDkuMzUxNDkgOS4xNjM0OEM5LjM0OTQ5IDkuMDUwOTggOS4zODk0OSA4Ljk0ODk4IDkuNDY0OTkgOC44NzE0OEM5LjUzODk5IDguNzk1NDggOS42Mzg5OSA4Ljc1MjQ4IDkuNzQ1NDkgOC43NTA0OEg5Ljc1MjQ5QzkuOTY4OTkgOC43NTA0OCAxMC4xNDg1IDguOTI0OTggMTAuMTUzNSA5LjE0MTk4QzEwLjE1MzUgOS4xNDE5OCAxMC4xNjUgOS4zNDk0OCAxMC4zMjA1IDkuNDUxOThDMTAuNTAyNSA5LjU3MTk4IDEwLjU1NDUgOS44MTc0OCAxMC40MzY1IDEwLjAwMVpNMTIuMTQ0IDUuMzE4QzEyLjAyODUgNC4zNDEgMTEuNTk2IDMuNDM4IDEwLjkwMTUgMi43NEMxMC4xMDUgMS45Mzk1IDkuMDUyNSAxLjUgNy45Mzg1IDEuNUM2LjgyNDUgMS41IDUuNzcyNSAxLjkzOTUgNC45NzYgMi43NDA1QzQuMjc4NSAzLjQ0MSAzLjg0NSA0LjM0OSAzLjczMiA1LjMzMTVDMi43NzU1IDUuNDQ3NSAxLjg5MSA1Ljg5MyAxLjIwOCA2LjYwODVDMC40MjkgNy40MjYgMCA4LjUwNjUgMCA5LjY1MDVDMCAxMi4wNDkgMS45IDE0IDQuMjM2NSAxNEgxMS43NjNDMTQuMDk4NSAxNCAxNiAxMi4wNDkgMTYgOS42NTA1QzE2IDcuMzg0IDE0LjMwMiA1LjUxNjUgMTIuMTQ0IDUuMzE4WiIgZmlsbD0iI0ZGRkZGRiIgLz4KPC9nPgo8L3N2Zz4K'

  return `
<!DOCTYPE html>
<html lang="${locale}">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#fff">
    <style type="text/css">${fontsCss(instance)}</style>
    <style type="text/css">${cozyBsCss}</style>
    <style type="text/css">${themeCss}</style>
    <style type="text/css">${cirrusCss}</style>
  </head>
  <body class="theme-inverted">
    <form id="login-form" method="POST" action="#" class="d-contents">
      <input id="trusted-device-token" type="hidden" name="trusted-device-token" value="" />
      <main class="wrapper">

        <header class="wrapper-top">
          <a href="#" id="backButton" class="btn btn-icon" aria-label="${strBackButton}">
            <span class="icon icon-back"></span>
          </a>
        </header>

        <div class="d-flex flex-column align-items-center">
          <img src="${avatarSvgUrl}" alt="" id="avatar" class="avatar my-3 border border-primary border-2 rounded-circle" style="box-sizing: content-box;" />
          <h1 class="h4 h2-md mb-0 text-center">${encode(title)}</h1>
          <p class="mb-4 mb-md-5 text-muted">${encode(fqdn)}</p>
          <div id="login-field" class="input-group form-floating has-validation w-100">
            <input type="password" class="form-control form-control-md-lg" id="password" name="passphrase" autofocus autocomplete="current-password" />
            <label for="password">${strPasswordField}</label>
            <button id="password-visibility-button" class="btn btn-outline-info"
              type="button"
              name="password-visibility"
              data-show="${strPasswordShow}"
              data-hide="${strPasswordHide}"
              title="${strPasswordShow}">
              <span id="password-visibility-icon" class="icon icon-eye-closed"></span>
            </button>
          </div>
          <a href="#" id="forgotPassLink" class="align-self-start my-3">
            ${strForgotPassword}
          </a>
        </div>

        <footer class="w-100">
          <button id="login-submit" class="btn btn-primary btn-md-lg w-100 my-3 mt-md-5" type="submit">
            ${strSubmit}
          </button>
        </footer>

      </main>
    </form>
    <script>${cirrusJs}</script>
    <script>${passwordHelperJs}</script>
    <script>${passwordVisibilityJs}</script>
    <script>${loginJs}</script>
    <script>
      function sendLoadedEvent() {
        setTimeout(() => {
          const avatarImg = document.getElementById('avatar')
          const avatarPosition = avatarImg.getBoundingClientRect()
          const height = avatarImg.clientHeight
          const width = avatarImg.clientWidth

          postMessage(JSON.stringify({
            message: 'loaded',
            avatarPosition: {
              top: avatarPosition.top,
              left: avatarPosition.left,
              height: height,
              width: width,
              boxHeight: avatarPosition.height,
              boxWidth: avatarPosition.width,
            }
          }))
        }, 100)
      }

      window.addEventListener("load", function(event) {
        var img = document.getElementById('avatar')

        if (img.complete) {
          sendLoadedEvent()
        } else {
          img.addEventListener('load', sendLoadedEvent)
          img.addEventListener('error', function() {
            alert('error')
          })
        }

        document.getElementById('forgotPassLink').onclick = () => {
          postMessage(JSON.stringify({
            message: 'forgotPassword'
          }))

          return false
        }

        document.getElementById('backButton').onclick = () => {
          postMessage(JSON.stringify({
            message: 'backButton'
          }))

          return false
        }
      })

      ${handleErrorMessage()}
      ${readonlyJs(['login-submit', 'password'])}
    </script>
  </body>
</html>
`
}
