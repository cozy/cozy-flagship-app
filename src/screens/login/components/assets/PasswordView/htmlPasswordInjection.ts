/********************************************************************/
/* This code should reflect cozy-stack/assets/templates/login.html */
/********************************************************************/

import { encode } from 'html-entities'

import { handleErrorMessage } from './js/jsHandleErrorMessage'
import { loginJs } from './js/jsLogin'
import { passwordHelperJs } from './js/jsPasswordHelper'
import { passwordVisibilityJs } from './js/jsPasswordVisibility'
import { cirrusCss } from '../common/css/cssCirrus'
import { cozyBsCss } from '../common/css/cssCozyBs'
import { fontsCss } from '../common/css/cssFonts'
import { themeCss } from '../common/css/cssTheme'
import { cirrusJs } from '../common/js/jsCirrus'
import { readonlyJs } from '../common/js/jsReadonly'

import { getDimensions } from '/libs/dimensions'
import type { ClouderyTheme } from '/screens/login/components/functions/clouderyThemeFetcher'
import { cssPadding } from '/screens/login/components/functions/webViewPaddingInjection'
import { getAvatarDataUri } from '/ui/Logo/avatar'

const strBackButton = "Revenir Ã  l'Ã©cran prÃ©cÃ©dent"
const strPasswordField = 'Mot de passe'
const strPasswordHide = 'Cacher le mot de passe'
const strPasswordShow = 'Afficher le mot de passe'
const strForgotPassword = "J'ai oubliÃ© mon mot de passe"
const strSubmit = 'SE CONNECTER'

const locale = 'fr'

console.log('ðŸ’œ getDimensions() from PasswordView')
const dimensions = getDimensions()

const getCustomThemeLink = (clouderyTheme: ClouderyTheme): string =>
  clouderyTheme.themeUrl
    ? `<link rel="stylesheet" media="screen" href="${clouderyTheme.themeUrl}">`
    : ''

/**
 * Generate password specific HTML to provide to webview
 *
 * @param {string} title - The title of the page - for example the user's name as configured in the Cozy's settings
 * @param {string} fqdn - The subtitle of the page - for example the Cozy's fqdn
 * @param {string} instance - The Cozy's url, used to get avatar and fonts css
 * @param {string} backgroundColor - The LoginScreen's background color (used for overlay and navigation bars)
 * @returns {string} HTML of Password form to inject inside Webview
 */
export const getHtml = (
  title: string,
  fqdn: string,
  instance: string,
  clouderyTheme: ClouderyTheme
): string => {
  const avatarUrl = new URL(instance)
  avatarUrl.pathname = 'public/avatar'

  const avatarSvgUrl = getAvatarDataUri()

  return `
<!DOCTYPE html>
<html lang="${locale}">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="theme-color" content="#fff">
    <style type="text/css">${fontsCss(instance)}</style>
    <style type="text/css">${cozyBsCss}</style>
    <style type="text/css">${themeCss}</style>
    <style type="text/css">${cirrusCss}</style>
    <style type="text/css">${cssPadding}</style>
    ${getCustomThemeLink(clouderyTheme)}
  </head>
  <body class="cirrus theme-inverted">
    <form id="login-form" method="POST" action="#" class="d-contents">
      <input id="trusted-device-token" type="hidden" name="trusted-device-token" value="" />
      <main class="wrapper">

        <header class="wrapper-top">
          <a href="#" id="backButton" class="btn btn-icon" aria-label="${strBackButton}">
            <span class="icon icon-back"></span>
          </a>
        </header>

        <div class="d-flex flex-column align-items-center">
          <img src="${avatarSvgUrl}" alt="" id="avatar" class="avatar my-3 border border-primary border-2 rounded-circle" style="box-sizing: content-box;" />
          <h1 class="h4 h2-md mb-0 text-center">${encode(title)}</h1>
          <p class="mb-4 mb-md-5 text-muted">${encode(fqdn)}</p>
          <div id="login-field" class="input-group form-floating has-validation w-100">
            <input type="password" class="form-control form-control-md-lg" id="password" name="passphrase" autofocus autocomplete="current-password" />
            <label for="password">${strPasswordField}</label>
            <button id="password-visibility-button" class="btn btn-outline-info"
              type="button"
              name="password-visibility"
              data-show="${strPasswordShow}"
              data-hide="${strPasswordHide}"
              title="${strPasswordShow}">
              <span id="password-visibility-icon" class="icon icon-eye-closed"></span>
            </button>
          </div>
          <a href="#" id="forgotPassLink" class="align-self-start my-3">
            ${strForgotPassword}
          </a>
        </div>

        <footer class="w-100">
          <button id="login-submit" class="btn btn-primary btn-md-lg w-100 my-3 mt-md-5" type="submit">
            ${strSubmit}
          </button>
        </footer>

      </main>
    </form>
    <script>${cirrusJs}</script>
    <script>${passwordHelperJs}</script>
    <script>${passwordVisibilityJs}</script>
    <script>${loginJs}</script>
    <script>
      function sendLoadedEvent() {
        setTimeout(() => {
          const avatarImg = document.getElementById('avatar')
          const avatarPosition = avatarImg.getBoundingClientRect()
          const height = avatarImg.clientHeight
          const width = avatarImg.clientWidth

          postMessage(JSON.stringify({
            message: 'loaded',
            avatarPosition: {
              top: avatarPosition.top - ${dimensions.statusBarHeight},
              left: avatarPosition.left,
              height: height,
              width: width,
              boxHeight: avatarPosition.height,
              boxWidth: avatarPosition.width,
            }
          }))
        }, 100)
      }

      window.addEventListener("load", function(event) {
        var img = document.getElementById('avatar')

        if (img.complete) {
          sendLoadedEvent()
        } else {
          img.addEventListener('load', sendLoadedEvent)
          img.addEventListener('error', function() {
            alert('error')
          })
        }

        document.getElementById('forgotPassLink').onclick = () => {
          postMessage(JSON.stringify({
            message: 'forgotPassword'
          }))

          return false
        }

        document.getElementById('backButton').onclick = () => {
          postMessage(JSON.stringify({
            message: 'backButton'
          }))

          return false
        }
      })

      ${handleErrorMessage()}
      ${readonlyJs(['login-submit', 'password'])}
    </script>
  </body>
</html>
`
}
