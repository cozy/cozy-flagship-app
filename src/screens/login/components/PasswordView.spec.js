import { render } from '@testing-library/react-native'
import React from 'react'

import { PasswordView } from './PasswordView'

const mockSpy = jest.fn()

jest.mock('cozy-client', () => ({
  useClient: jest.fn().mockReturnValue({})
}))
jest.mock('@react-native-cookies/cookies', () => ({
  set: jest.fn()
}))
jest.mock('react-native-webview', () => {
  // eslint-disable-next-line @typescript-eslint/no-shadow
  const React = require('react')
  class WebView extends React.Component {
    postMessage(payload) {
      mockSpy(payload)
    }
    render() {
      return <div>WebView</div>
    }
  }
  return { WebView }
})

describe('PasswordView', () => {
  beforeEach(() => {
    jest.resetAllMocks()
  })

  it('should render component with webview and view', () => {
    const props = {
      instance: 'http://cozy.192-168-1-102.nip.io:8080',
      clouderyTheme: {
        backgroundColor: '#4b4b4b',
        themeUrl: 'SOME8URL'
      }
    }
    // When
    const { toJSON } = render(<PasswordView {...props} />)

    // Then
    expect(toJSON()).toMatchSnapshot()
  })

  describe('on error message', () => {
    it('should post error message', () => {
      const errorMessage = 'wrong password'
      const props = {
        instance: 'http://cozy.192-168-1-102.nip.io:8080',
        clouderyTheme: {
          backgroundColor: '#4b4b4b',
          themeUrl: 'SOME8URL'
        }
      }
      expect(mockSpy).toHaveBeenCalledTimes(0)

      const { update } = render(<PasswordView {...props} />)

      // When
      update(<PasswordView {...props} errorMessage={errorMessage} />)

      // Then
      expect(mockSpy).toHaveBeenCalledTimes(2)
      expect(mockSpy).toHaveBeenCalledWith(
        JSON.stringify({ message: 'setReadonly' })
      )
      expect(mockSpy).toHaveBeenNthCalledWith(
        2,
        JSON.stringify({ message: 'setErrorMessage', param: errorMessage })
      )
    })
  })
})
