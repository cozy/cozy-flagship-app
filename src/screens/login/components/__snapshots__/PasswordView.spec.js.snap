// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PasswordView should render component with webview and view 1`] = `
<View
  behavior="height"
  style={
    {
      "flex": 1,
    }
  }
>
  <div>
    WebView
  </div>
  <View
    style={
      [
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        {
          "backgroundColor": "#4b4b4b",
        },
      ]
    }
  />
</View>
`;
